### PR TITLE
LCOM4 metric implementation

### PR DESCRIPTION
**Pull request by silverCase**
#182

### CODE REVIEW

File: aibolit/metrics/lcom4/lcom4.py [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+                graph[node.name].add(this_m.selectors[0].member)
+
+            for _, mi in node.filter(MethodInvocation):
+                if mi.member in methods:
+                    graph[node.name].add(mi.member)
+
+        return self.get_connected_components(methods, fields, graph)
+
+    @staticmethod
+    def get_connected_components(methods, fields, graph):
```
Comment (acheshkov) [inline aibolit/metrics/lcom4/lcom4.py:69]: specify method return type

Comment (acheshkov) [reply aibolit/metrics/lcom4/lcom4.py:69]: and typings for arguments

**Comment by 0crat** [comment]
This pull request [#218](https://github.com/yegor256/aibolit/issues/218) is assigned to @lyriccoder[/z](https://www.0crat.com/u/lyriccoder), here is [why](https://www.0crat.com/footprint/CT5UB7XAP/e6b4cc5b-57eb-4d1a-955e-e1a43e169663); the budget is 15 minutes, see [§4](http://www.zerocracy.com/policy.html#4); please, read [§27](http://www.zerocracy.com/policy.html#27) and when you decide to accept the changes, inform @acheshkov[/z](https://www.0crat.com/u/acheshkov) (the architect) right in this ticket; if you decide that this PR should not be accepted ever, also inform the architect; this [blog post](http://www.yegor256.com/2015/02/09/serious-code-reviewer.html) will help you understand what is expected from a code reviewer; there will be no monetary reward for this job

File: aibolit/metrics/lcom4/lcom4.py (hunk 1 above)
Comment (silverCase) [reply aibolit/metrics/lcom4/lcom4.py]: Done!

File: aibolit/metrics/lcom4/lcom4.py [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+    MemberReference, FieldDeclaration, MethodInvocation, This
+
+
+class LCOM4:
+
+    def value(self, filename: str) -> int:
+        tree = AST(filename).value()
+        graph: Dict[str, Set[str]] = defaultdict(set)
+
+        fields = [node.declarators[0].name for _, node in tree.filter(FieldDeclaration)]
```
Comment (acheshkov) [inline aibolit/metrics/lcom4/lcom4.py:38]: can we add typings for every variable?

File: aibolit/metrics/lcom4/lcom4.py [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+
+        for i in methods:
+            G.add_node(i)
+        for i in fields:
+            G.add_node(i)
+        for key, val in graph.items():
+            for x in val:
+                G.add_edge(key, x)
+
+        # nx.draw(G, with_labels=True)
```
Comment (acheshkov) [inline aibolit/metrics/lcom4/lcom4.py:80]: remove commented code

Comment (silverCase) [reply aibolit/metrics/lcom4/lcom4.py]: Done

File: aibolit/metrics/lcom4/lcom4.py (hunk 2 above)
Comment (silverCase) [reply aibolit/metrics/lcom4/lcom4.py]: Done!

**Comment by silverCase** [comment]
#### For now, the algorithm only works fine in java classes

- [ ] with no attribute name shadowing 

- [ ] with no no subclasses. 

#### Therefore it needs more work.

**Comment by acheshkov** [comment]
@rultor merge

**Comment by 0crat** [comment]
Job was finished in 23 hours, bonus for fast delivery is possible (see [§36](http://www.zerocracy.com/policy.html#36))

**Comment by 0crat** [comment]
Job audit: performer @lyriccoder[/z](https://www.0crat.com/u/lyriccoder)[/z](https://www.0crat.com/u/lyriccoder) didn't make CR comments

**Pull request merged**